### PR TITLE
Amrsw 274 inflation dead lock

### PR DESCRIPTION
**Pull request by remmaTech12**
related to [AMRSW-274](https://lexxpluss.atlassian.net/browse/AMRSW-274)
explained in a jira ticket in detail

**Tests for VariableInflationLayer**
- in simulation
  - [x] variable inflation works when velocity changes
  - [x] in configuration yaml change,
    - [x] min_inflation_radius changes
    - [x] max_inflation_radius changes
    - [x] min_inflation_vel changes
    - [x] max_inflation_vel changes 
  - [x] 100 times testing without dead-lock
  - [x] run for at least 15 mins in the case that an obstacle is in the way
 
- with robot
  - [x] 100 times testing without dead-lock
  - [x] run for at least 5 mins in the case that an obstacle is in the way

**Regression Tests for InflationLayer**
- in simulation
  - [x] inflation layer works normally (inflation appears
  - [x] run for at least 15 mins in the case that an obstacle is in the way
- with robot
  - [x] inflation layer works normally (inflation appears
  - [x] run for at least 5 mins in the case that an obstacle is in the way

[AMRSW-274]: https://lexxpluss.atlassian.net/browse/AMRSW-274?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by remmaTech12** [comment]
InflationLayer を継承して VariableInflationLayer を作ろう（和田さんのアドバイス）と思ったが、
- ros2 に移行するときに navigation package は大規模に変更するかもしれない
- InflationLayer は今後変更しないと思われる

ので、継承はやめておいた。

File: costmap_2d/include/costmap_2d/variable_inflation_layer.h [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+protected:
+  virtual void onFootprintChanged();
+  boost::recursive_mutex* inflation_access_;
+  boost::recursive_mutex* velocity_access_;
+
+  double resolution_;
+  double inflation_radius_;
+  double inscribed_radius_;
+  double weight_;
+  bool inflate_unknown_;
```
Comment (remmaTech12) [inline costmap_2d/include/costmap_2d/variable_inflation_layer.h:151]: 基本的に inflation を可変にするので、use_variable_inflation_ というパラメータは削除した。

File: costmap_2d/plugins/variable_inflation_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+  , last_min_x_(-std::numeric_limits<float>::max())
+  , last_min_y_(-std::numeric_limits<float>::max())
+  , last_max_x_(std::numeric_limits<float>::max())
+  , last_max_y_(std::numeric_limits<float>::max())
+{
+  inflation_access_ = new boost::recursive_mutex();
+  velocity_access_ = new boost::recursive_mutex();
+
+  ros::NodeHandle nh("~/" + name_), g_nh;
+  diff_drive_debug_info_sub = g_nh.subscribe<lexxauto_msgs::DiffDriveEffortControllerDebug>
```
Comment (remmaTech12) [inline costmap_2d/plugins/variable_inflation_layer.cpp:73]: onInitialize() で subscriber を定義すると、dead lock が発生する問題があったためコンストラクタで定義。

File: costmap_2d/plugins/variable_inflation_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+            " inscribed_radius_ = %.3f, inflation_radius_ = %.3f",
+            layered_costmap_->getFootprint().size(), inscribed_radius_, inflation_radius_);
+}
+
+double VariableInflationLayer::calculate_variable_inflation_radius()
+{
+  double measured_velocity = 0.0;
+  double variable_inflation_radius = min_inflation_radius_;
+  {
+    boost::unique_lock < boost::recursive_mutex > lock(*velocity_access_);
```
Comment (remmaTech12) [inline costmap_2d/plugins/variable_inflation_layer.cpp:191]: mutex_lock をかける範囲を小さくした（以前は関数全体にかけていた）

**Review by kokosabu — APPROVED** [review]

**Review by yakato-jun — APPROVED** [review]

**Pull request merged**